### PR TITLE
Tuya TS0601 Temp Humidity with Screen (_TZE200_bjawzodf)

### PR DESCRIPTION
**Pull request by jenko1080**
Sensor: https://www.aliexpress.com/item/4000739457722.html
Unable to find any payload for RSSI, only Battery / Temp / Humidity seems to be sent through.

Happy for any suggested changes / optimisations, not sure how much else of the magic inside the device works.

Tested variable battery voltages.
image

Fixes: #1561

### CODE REVIEW

**Comment by javicalle** [comment]
Already reported here:
* https://github.com/zigpy/zha-device-handlers/issues/1561

That is my proposal (untested):
<details>
<summary>ts0601_sensor.py</summary>

```yaml
"""Tuya Temperature, Humidity sensors."""

from typing import Dict

from zigpy.profiles import zha
from zigpy.profiles.zha import DeviceType
from zigpy.quirks import CustomDevice
from zigpy.zcl.clusters.general import Basic, Groups, Ota, Scenes, Time
from zigpy.zcl.clusters.measurement import RelativeHumidity, TemperatureMeasurement

from zhaquirks.const import (
    DEVICE_TYPE,
    ENDPOINTS,
    INPUT_CLUSTERS,
    MODELS_INFO,
    OUTPUT_CLUSTERS,
    PROFILE_ID,
)
from zhaquirks.tuya import TuyaLocalCluster
from zhaquirks.tuya.mcu import (
    DPToAttributeMapping,
    TuyaDPType,
    TuyaMCUCluster,
    TuyaPowerConfigurationCluster,
)


class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
    """Tuya local RelativeHumidity cluster."""


class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
    """Tuya local TemperatureMeasurement cluster."""


class TemperatureHumidityManufCluster(TuyaMCUCluster):
    """Tuya Manufacturer Cluster with Temperature and Humidity data points."""

    dp_to_attribute: Dict[int, DPToAttributeMapping] = {
        1: DPToAttributeMapping(
            TuyaTemperatureMeasurement.ep_attribute,
            "measured_value",
            dp_type=TuyaDPType.VALUE,
        ),
        2: DPToAttributeMapping(
            TuyaRelativeHumidity.ep_attribute,
            "measured_value",
            dp_type=TuyaDPType.VALUE,
        ),
        4: DPToAttributeMapping(
            TuyaPowerConfigurationCluster.ep_attribute,
            "battery_percentage_remaining",
            dp_type=TuyaDPType.VALUE,
        ),
    }

    data_point_handlers = {
        1: "_dp_2_attr_update",
        2: "_dp_2_attr_update",
        4: "_dp_2_attr_update",
    }


class TemperatureHumidtySensor(CustomDevice):
    """Tuya Temperature and Humidity sensor device."""

    signature = {
        MODELS_INFO: [("_TZE200_bjawzodf", "TS0601")],
        ENDPOINTS: {
            # <SimpleDescriptor endpoint=1, profile=260, device_type=81
            # device_version=1
            # input_clusters=[0, 4, 5, 61184]
            # output_clusters=[25, 10]>
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.SMART_PLUG,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    TemperatureHumidityManufCluster.cluster_id,
                ],
                OUTPUT_CLUSTERS: [Ota.cluster_id, Time.cluster_id],
            }
        },
    }

    replacement = {
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: DeviceType.TEMPERATURE_SENSOR,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    TemperatureHumidityManufCluster,
                    TuyaTemperatureMeasurement,
                    TuyaRelativeHumidity,
                    TuyaPowerConfigurationCluster,
                ],
            },
        },
    }
```

</details>

**Comment by jpillora** [comment]
My implementation was the first thing that worked, my python and experience with zigpy is lacking so I’d be happy to use a more standard implementation. Thank you for PR, I’ll try yours out today :)

**Comment by jenko1080** [comment]
hey @javicalle , much nicer code! I'm on a similar page as @jpillora being a bit weak with Python.

The base I was using to reference for the implementation was [zhaquirks/tuya/ts0601_siren.py](https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/ts0601_siren.py#L88), it looks like @jpillora had found the same place. I started an earlier discussion in the wrong place 4 months ago [here](https://community.home-assistant.io/t/temperature-humidity-sensor-zha-configuration-tze200-bjawzodf-ts0601/397095) and didn't notice the responses months later!

@jpillora, for the battery voltage fix to our common code, I've included `from zhaquirks.ikea import PowerConfiguration2AAACluster`, that has the appropriate settings to tell the device it's a battery among other things. It didn't seem sensible to copy / paste this code from `ikea` to `tuya`- though from a maintainability point this could be really irritating in future.

@javicalle, do you know if anyone has a strong opinion on where the 2AAA / 3AAA classes should live?
- [zhaquirks/ikea/__init__.py](https://github.com/zigpy/zha-device-handlers/blob/b6556169bf641e8f2eb882e8ace4a5a1ab4552d5/zhaquirks/ikea/__init__.py#L73)
- [zhaquirks/tuya/__init__.py](https://github.com/zigpy/zha-device-handlers/blob/b6556169bf641e8f2eb882e8ace4a5a1ab4552d5/zhaquirks/tuya/__init__.py#L807)

I'll add an update to my PR to include 2AAA in the tuya init anticipating a 'keep it all together' themed response for that question.

**Comment by javicalle** [comment]
They are classes with different hierarchies, so I don't think they can be used in the same way.
In any case, if I wanted to extract the common factor to a generic class, I would surely put it in `zhaquirk.__init__.py` 🤷🏻‍♂️

**Comment by mj0c** [comment]
Perfect, already done in the last PR :)
Final step is to jump across to the neater `dp_to_attribute` implementation, but I'm out of time for a few days with work.

I saw a few possibilities to get the RSSI but not sure how to validate them without guess/check. Are there any common methods anyone uses to find magic sequence to poll the device?

**Comment by javicalle** [comment]
> I saw a few possibilities to get the RSSI but not sure how to validate them without guess/check. Are there any common methods anyone uses to find magic sequence to poll the device?

Sorry but I'm not familiar with that part.
Isn't that something that is handled automagically at the zigpy API level?

**Comment by jenko1080** [comment]
I believe it should be, but no data appears to flow. Only the 3x pieces of information (temp, hum, batt) are offered currently.

There is the mention of a Tuya spell in the issue you linked: https://github.com/zigpy/zha-device-handlers/issues/1561#issuecomment-1125070832.

But without getting a Tuya hub and soldering on some probes to see how it all works I'm not sure how to request the data from the device... if it's even available.

**Comment by javicalle** [comment]
I'm pretty sure that the `lqi` and `RSSI` aren't part of the MCU messaging that you are checking.

**Comment by MattWestb** [comment]
All tuya TH(L) with screen is known being tricky and and need tuya magic spell being casted on them for working OK and reporting all things OK so also very likely it the reason way not getting RSSI from the device (its stored in the frame the device is sending and i think its calculated from the last sent frame(s) from its parent) so its 100 % the Zigbee SOC that is doing it wrong and not the tuya MCU.

**Comment by jenko1080** [comment]
@MattWestb , how did you find the last Tuya spell you wrote?
And, because I'm definitely new to this I'm going to ask the stupid follow-up question: once that spell is added have the other devices you wrote spells for sent the data through on the Basic cluster?

**Comment by MattWestb** [comment]
Some Z2M user was sniffing one TS004F dimmer switch then paring it with tuya ZBGW and i was analyzing it and one Hubitar dev was testing it and its working and they is using it for many tuya device and Z2M is also doing it. 
The code for TS004F:
https://github.com/zigpy/zha-device-handlers/blob/b6556169bf641e8f2eb882e8ace4a5a1ab4552d5/zhaquirks/tuya/ts004f.py#L187-L199.

Its also one version is being used for LIDL / tuya power strips that dont need it then paring for getting all endpoints working OK.

We dont knowing how its working inside tuya devices but it can changing the hardware and software setting of the device and how its acting in the system and its real magic !!

**Comment by javicalle** [comment]
> @MattWestb , how did you find the last Tuya spell you wrote?

You can import the `EnchantedDevice` and extend your device from it.
* add the import:
```python
from zhaquirks.tuya.mcu import EnchantedDevice
```
* extend your class:
```python
class TuyaTempHumidity(EnchantedDevice):
```

I have updated my version to the Tuya Spell Magic

**Comment by jpillora** [comment]
I figured we don’t need the spell because we’re getting the values already,
but I could be wrong?

**Comment by jenko1080** [comment]
There's at least 1 more metric visible on the screen, being the RSSI or LQI. I feel that if it's displayed, surely it must be available.

I'm willing to bet that using the Tuya Smart Life app and a sniffer with a ZB Bridge, you should be able to see what the app sends to the device, and then copy that magic string in to the python code to get it started... It may also be an on-demand metric, but still sniffable.

File: zhaquirks/tuya/__init__.py [hunk 1]
```diff
@@ -1313,6 +1313,20 @@ def update_attribute(self, attr_name: str, value: Any) -> None:
         return self._update_attribute(attr.id, value)
 
 
+class TuyaPowerConfigurationCluster2AAA(PowerConfiguration, TuyaLocalCluster):
+    """PowerConfiguration cluster for battery-operated TRVs with 2 AAA."""
```
Comment (javicalle) [inline zhaquirks/tuya/__init__.py:1317]:
```python
"""PowerConfiguration cluster for devices with 2 AAA."""
```

File: zhaquirks/tuya/ts0601_temperaturehumidity.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    attributes = TuyaManufClusterAttributes.attributes.copy()
+    attributes.update(
+        {
+            TUYA_TEMPERATURE_ATTR: ("temperature", t.uint32_t, True),
+            TUYA_HUMIDITY_ATTR: ("humidity", t.uint32_t, True),
+            TUYA_BATTERY_ATTR: ("battery", t.uint8_t, True),
+        }
+    )
+
```
Comment (javicalle) [inline zhaquirks/tuya/ts0601_temperaturehumidity.py:63]: No need to keep as cluster attributes.

File: zhaquirks/tuya/ts0601_temperaturehumidity.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(self, *args, **kwargs):
+        """Init device."""
+
+        self.temperature_bus = Bus()
+        self.humidity_bus = Bus()
+        self.battery_bus = Bus()
+
+        super().__init__(*args, **kwargs)
+
```
Comment (javicalle) [inline zhaquirks/tuya/ts0601_temperaturehumidity.py:103]: No need for attributes Bus since we are not triggering events.

**Review by javicalle — CHANGES_REQUESTED** [review]
Some remanent from previous implementation can be removed.

File: zhaquirks/tuya/__init__.py (hunk 1 above)
Comment (jenko1080) [reply zhaquirks/tuya/__init__.py]: Reworded (what did TRV stand for?)

File: zhaquirks/tuya/ts0601_temperaturehumidity.py (hunk 2 above)
Comment (jenko1080) [reply zhaquirks/tuya/ts0601_temperaturehumidity.py]: Removed, and unnecessary constants.

File: zhaquirks/tuya/ts0601_temperaturehumidity.py (hunk 3 above)
Comment (jenko1080) [reply zhaquirks/tuya/ts0601_temperaturehumidity.py]: Removed.

File: zhaquirks/tuya/ts0601_temperaturehumidity.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
+    """Tuya local TemperatureMeasurement cluster."""
+
+
+class TuyaRelativeHumidity(RelativeHumidity, TuyaLocalCluster):
+    """Tuya local RelativeHumidity cluster."""
+
+
+class TemperatureHumidityManufCluster(TuyaNewManufCluster):
```
Comment (javicalle) [inline zhaquirks/tuya/ts0601_temperaturehumidity.py:44]: Please extend from `zhaquirks.tuya.mcu.TuyaMCUCluster`.
I have plans to migrate all `TuyaNewManufCluster` to `TuyaMCUCluster` in the long term.

File: zhaquirks/tuya/__init__.py [hunk 5]
```diff
@@ -1313,6 +1313,20 @@ def update_attribute(self, attr_name: str, value: Any) -> None:
...
+    BATTERY_QUANTITY = 0x0033
+    BATTERY_RATED_VOLTAGE = 0x0034
+
+    _CONSTANT_ATTRIBUTES = {
+        BATTERY_SIZES: 4,
+        BATTERY_QUANTITY: 2,
+        BATTERY_RATED_VOLTAGE: 15,
+    }
+
+
```
Comment (javicalle) [inline zhaquirks/tuya/__init__.py:1329]: Please, put the class definition next to `TuyaPowerConfigurationCluster2AA` and `TuyaPowerConfigurationCluster3AA` classes.

**Review by javicalle — CHANGES_REQUESTED** [review]
A few more changes.

Also, not sure about the `ts0601_temperaturehumidity.py` filename. Maybe new sensors can be added to the file in a future, eg. Temp&Hum&Illum. But not sure about the proper name. `ts0601_sensor.py`?

File: zhaquirks/tuya/ts0601_temperaturehumidity.py (hunk 4 above)
Comment (jenko1080) [reply zhaquirks/tuya/ts0601_temperaturehumidity.py:44]: done

File: zhaquirks/tuya/__init__.py (hunk 5 above)
Comment (jenko1080) [reply zhaquirks/tuya/__init__.py]: Moved up with dependency `TuyaLocalCluster`. 
------

The power clusters inherited from different bases
`class TuyaPowerConfigurationCluster2AA(TuyaPowerConfigurationCluster)`
and
`class TuyaPowerConfigurationCluster2AAA(PowerConfiguration, TuyaLocalCluster)`

Added some minor refactoring to bring `TuyaPowerConfigurationCluster` across to inheriting from `TuyaLocalCluster` (it already used attribute name to get the id)

**Review by javicalle — APPROVED** [review]
LGTM

**Comment by javicalle** [comment]
@mj0c please, can you add the `Fixes` mark in your first comment (just add this line at the end):
```
Fixes: #1561
```

This way, the issue will be closed once the PR is merged.

Thanks in advanced.

**Comment by jenko1080** [comment]
done @javicalle , thanks for all your support on this.

**Comment by jpillora** [comment]
Thanks @jenko1080 !

**Pull request merged**